### PR TITLE
feature: add asset on TicketRecurrent to add them on the ticket creat…

### PR DESCRIPTION
**Pull request by delcroip**
replace #11127 (avoid too many commits)

new version https://github.com/glpi-project/glpi/pull/9567 that is using GLPI 10 file names and structures

Task to be done:

    migrate methods from ticket to commonItilObject : done
    updating the Item_TicketRecurrent/ TicketRecurrent to add devices support: Done
    adding the strategy setting (1 ticket for all or 1 ticket per device) in TicketRecurent: done
    adding the device in the created ticket according to strategy: Done
    manual testing strategy 1 item per ticket: passed
    manual testing all item on 1 ticket: passed
    adding tests : TBD

NB: ChangeRecurrent won't be done in that PR

### CODE REVIEW

**Review by orthagh — COMMENTED** [review]
functional review : OK
technical review : still in progress (the code displacement doesn't help the review)

File: ajax/commonitilobjectiteminformation.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+// Direct access to file
+
```
Comment (orthagh) [inline ajax/commonitilobjectiteminformation.php:35]: maybe you can check the existence (and the inheritance) of obj_class and item_obj_class and die if they don't exist ?

Also is item_obj_class used ? I see the declaration in the 2 other ajax files but no usage

File: src/CommonITILObject.php [hunk 2]
```diff
@@ -6504,11 +6562,12 @@ public static function showShort($id, $options = [])
...
                // Ticket : simple link to item
+               // extend to other ITIL object ? new Item_Ticket(); --> new Item_${self::class}();
                 $sixth_col  = "";
                 $is_deleted = false;
-                $item_ticket = new Item_Ticket();
-                $data = $item_ticket->find(['tickets_id' => $item->fields['id']]);
-
+                $classname = "Item_" . static::class;
+                $item_ticket = new $classname();
+                $data = $item_ticket->find([strtolower(static::class) . 's_id' => $item->fields['id']]);
```
Comment (orthagh) [inline src/CommonITILObject.php:6570]: same here, use getForeignKeyFieldForItemType

File: src/CommonItilObject_Item.php [hunk 3]
```diff
@@ -36,11 +36,1321 @@
...
+    }
+
+    /**
+     * validate the object type is align with itemtype_1
+     *
+     * @param $obj  object holding the item
+     *     *
+     * @return bool
+     **/
+
```
Comment (orthagh) [inline src/CommonItilObject_Item.php:173]:
```suggestion
```

File: src/CommonItilObject_Item.php [hunk 4]
```diff
@@ -36,11 +36,1321 @@
...
+    /**
+     * Generate JavaScript Code to load the list of itmes
+     *
+     * @param $rand  int random number use for dropdown name
+     * @param $opt   array of possible options to be include in the call
+     *
+     * @return void
+     **/
+    public static function getJs($rand, $opt)
```
Comment (orthagh) [inline src/CommonItilObject_Item.php:346]: I'm not 100% sure, but this function seems unused, maybe we can deprecate it (@cedric-anne ?)

File: src/CommonItilObject_Item.php [hunk 5]
```diff
@@ -210,4 +1645,42 @@ public static function dropdownAllDevices(
         }
         return $rand;
     }
+
+    /** add items_2 on item_1 (fonction may already exist elswhere but I couldn't find it )
```
Comment (orthagh) [inline src/CommonItilObject_Item.php:1649]:
```suggestion
    /**
     * add items_2 on item_1 (fonction may already exist elswhere but I couldn't find it )
```

Comment (orthagh) [reply src/CommonItilObject_Item.php:1649]: you can clean the comment also

**Review by orthagh — CHANGES_REQUESTED** [review]

File: ajax/commonitilobjectiteminformation.php (hunk 1 above)
Comment (delcroip) [reply ajax/commonitilobjectiteminformation.php:35]: then I will remove it

File: src/CommonItilObject_Item.php (hunk 4 above)
Comment (delcroip) [reply src/CommonItilObject_Item.php:345]: Thanks, this one was required on 9.5,

the code in version 10 is there https://github.com/glpi-project/glpi/blob/4fb4226b8f8b3559939c1de89eec9bed63a4f41d/templates/components/itilobject/add_items.html.twig#L89

but its hardcoded to 'itemTicket'  I addapted the code

**Comment by orthagh** [comment]
Why drop `ajax/commonitilobjectiteminformation.php` in the a5c1ee12517c0f88011e88f604e887651c6e223b commit ?

relative to initial statement:

> NB: ChangeRecurrent won't be done in that PR

Note, this not ease review, i was lost for a moment to check the changes and see if we are in a merge-able state

**Comment by delcroip** [comment]
>
simply because it is not used in that PR, only ticketinformation is used (ticket recurent create a ticket instance), might be used by ChangeRecurent later.

Regarding the change recurrent, I though this PR was already big enough

File: src/CommonItilObject_Item.php [hunk 6]
```diff
@@ -38,11 +38,1292 @@
...
+            // Global search
+            ob_start();
+            static::dropdownAllDevices("itemtype", $params['itemtype'], 0, 1, $params['_users_id_requester'], $obj->fields["entities_id"], $p);
+            $twig_params['all_items_dropdown'] = ob_get_clean();
+
+            //echo "<span id='item_".strtolower(static::$itemtype_1)."_selection_information$rand'></span>";
+            //echo "</div>";
+
+            // Add button
+            //echo "<a href='javascript:itemAction$rand(\"add\");' class='vsubmit' style='float:left'>"._sx('button', 'Add')."</a>";
```
Comment (trasher) [inline src/CommonItilObject_Item.php:297]:
```suggestion
```

File: src/CommonItilObject_Item.php [hunk 7]
```diff
@@ -38,11 +38,1292 @@
...
+        }
+
+        // Display list
+        //echo "<div style='clear:both;'>";
+
```
Comment (trasher) [inline src/CommonItilObject_Item.php:302]:
```suggestion
```

**Comment by cedric-anne** [comment]
@delcroip 

I started the review of this PR and directly fixed some pieces of code to be able to merge it quickly (see https://github.com/cedric-anne/glpi/commit/3fecf3591ef99e8350b7d177ca49a6b275d1e1bf ).
Could I push my changes on your branch ? As I also rebased on current state of main branch, I could only do it using a force-push.

For now, I only made a technical review. I have to do some manual tests.

**Comment by cedric-anne** [comment]
Manually merged in 11f72d9c1acf37f48dbf48fe798ced57c2094d65.

**Comment by cedric-anne** [comment]
Thanks for the contribution.

**Pull request closed without merging**